doc/LIMITS.md

```diff
@@ -0,0 +1,251 @@
+  metrics: {
+    enabled: true
+  },
+  connectionManager: {
+    /**
+     * If the node transfers more than this amount of data in bytes/second
+     * low value connections may be closed
+     */
+    maxData: 1024 * 1024,
+
+    /**
+     * If the node sends more than this amount of data in bytes/second
+     * low value connections may be closed
+     */
+    maxSentData: 1024 * 1024
+
+    /**
+     * If the node receives more than this amount of data in bytes/second
+     * low value connections may be closed
+     */
+    maxReceivedData: 1024 * 1024,
+
+    /**
+     * If the event loop takes longer than this many ms to run,  low value
+     * connections may be closed
+     */
+    maxEventLoopDelay: 1000
+  }
+})
+```
+
+## Stream limits
+
+libp2p stream multiplexers impose limits on the amount of streams that can be opened per connection, and also the amount of data that will be buffered for a given stream. The data should be consumed as fast as possible - if a stream's input buffer exceeds the limits set the stream will be reset.
+
+These settings are done on a per-muxer basis, please see the README of the relevant muxer you are using.
+
+### Mplex
+
+[@libp2p/mplex](https://github.com/libp2p/js-libp2p-mplex) supports the following:
+
+```js
+const node = await createLibp2pNode({
+  muxers: [
+    new Mplex({
+      /**
+       * The total number of inbound protocol streams that can be opened on a given connection
+       */
+      maxInboundStreams: 1024,
+
+      /**
+       * The total number of outbound protocol streams that can be opened on a given connection
+       */
+      maxOutboundStreams: 1024,
+
+      /**
+       * How much incoming data to buffer before resetting the stream
+       */
+      maxStreamBufferSize: 4 * 1024 * 1024,
+
+      /**
+       * Mplex does not support backpressure so to protect ourselves, if `maxInboundStreams` is
+       * hit and the remote opens more than this many streams per second, close the connection
+       */
+      disconnectThreshold: 5
+    })
+  ]
+})
+```
+
+### Yamux
+
+[@chainsafe/libp2p-yamux](https://github.com/Chainsafe/js-libp2p-yamux) supports the following:
+
+```js
+const node = await createLibp2pNode({
+  muxers: [
+    new Yamux({
+      /**
+       * The total number of inbound protocol streams that can be opened on a given connection
+       */
+      maxInboundStreams: 1024,
+
+      /**
+       * The total number of outbound protocol streams that can be opened on a given connection
+       */
+      maxOutboundStreams: 1024
+    })
+  ]
+})
+```
+
+### Protocol limits
+
+When registering listeners for custom protocols, the maximum number of simultaneously open inbound and outbound streams per-connection can be specified. If not specified these will default to 32 inbound streams and 64 outbound streams.
+
+If more than this number of streams for the given protocol are opened on a single connection, subsequent new streams for that protocol will be immediately reset.
```

Comment:
Is there a mechanism to coalesce connections? What happens if peer B opens 10 connections to peer A. Does peer A see that there already is a connection to peer B and close the older ones?

I'm wondering because this limit is per connection, and I'm curious why it isn't per peer.

Reply:
Not yet, all we can do is just close duplicate connections I think, at least until the stream migration protocol gets a bit more fleshed out.